src/vdom/diff.js

```diff
@@ -152,15 +152,17 @@ function idiff(dom, vnode, context, mountAll) {
 	diffAttributes(out, vnode.attributes, props);
 
 
-	// Optimization: fast-path for elements containing a single TextNode:
-	if (!hydrating && vchildren && vchildren.length===1 && typeof vchildren[0]==='string' && fc && fc instanceof Text && !fc.nextSibling) {
-		if (fc.nodeValue!=vchildren[0]) {
-			fc.nodeValue = vchildren[0];
+	if (!props.dangerouslySetInnerHTML) {
+		// Optimization: fast-path for elements containing a single TextNode:
+		if (!hydrating && vchildren && vchildren.length===1 && typeof vchildren[0]==='string' && fc && fc instanceof Text && !fc.nextSibling) {
+			if (fc.nodeValue!=vchildren[0]) {
+				fc.nodeValue = vchildren[0];
+			}
+		}
+		// otherwise, if there are existing or new children, diff them:
+		else if (vchildren && vchildren.length || fc) {
```

Comment:
thinking since the first `if` block will never be true during hydration, this could actually just go into the second block:

```js
else if (!props.dangerouslySetInnerHTML && vchildren && vchildren.length || fc) {
```